reservation_units/utils/first_reservable_time_helper/first_reservable_time_helper.py

```diff
@@ -219,34 +269,90 @@ def calculate_all_first_reservable_times(self) -> None:
-        if self.pagination_args is not None:
-            self.pagination_args["size"] = chunk_size
-            qs_slice = calculate_queryset_slice(**self.pagination_args)
-            chunk_size = qs_slice.stop - qs_slice.start
-            qs = qs[qs_slice.start :]
+            # Otherwise, we can still skip looping through the previous pages.
+            qs = qs[self.start_offset :]
+            self.start_offset = 0  # Remove offset since we skip the previous pages
 
-        page_size: int = 0
-        # If we should only show reservable reservation units, increase the chunk size to max page size
-        # so that if filtering occurs, we don't need to fetch so many chunks.
-        size = graphene_settings.RELAY_CONNECTION_MAX_LIMIT if self.show_only_reservable else chunk_size
+        # If we don't have valid results, and this is not the first page,
+        # we should fetch the current and previous pages in on chunk. The next chunk
+        # will also be bigger (if needed), but likely small enough (<100) not to cause any problems.
+        elif self.start_offset > 0:
+            self.chunk_size = self.stop_offset
 
-        for reservation_unit in qs.hooked_iterator(self._get_elements_for_reservation_units, chunk_size=size):
+        results: int = 0
+        for reservation_unit in qs.hooked_iterator(self._get_affecting_time_spans, chunk_size=self.chunk_size):
             helper = ReservationUnitFirstReservableTimeHelper(parent=self, reservation_unit=reservation_unit)
             is_closed, first_reservable_time = helper.calculate_first_reservable_time()
 
             self.reservation_unit_closed_statuses[reservation_unit.pk] = is_closed
             self.first_reservable_times[reservation_unit.pk] = first_reservable_time
 
-            # If we have pagination args, we can stop if we have enough reservation units to fill the page size.
-            # If 'show_only_reservable' is True, we might need to fetch more than the first chunk size.
-            if self.pagination_args is not None:
-                if not self.show_only_reservable or first_reservable_time is not None:
-                    page_size += 1
+            # Start offset should exist only if we need to recalculate previous pages.
+            # -> Don't count the result for the current page.
+            if self.start_offset > 0:
+                self.start_offset -= 1
+                continue
+
+            # If we should only show reservable reservation units, then we should count the result for the
+            # current page only if there is a first reservable time.
+            if not self.show_only_reservable or first_reservable_time is not None:
+                results += 1
+
+            # This only really matters when showing only reservable reservation units,
+            # since we might need to fetch more than the first chunk size.
+            if results >= self.needed_reservation_units:
+                break
+
+        self._cache_results()
```

Comment:
Does this not effectively extend the caching period for the ru.pks (of the previously queried pages) that were retrieved from the cache during the processing of this request -- so that if there are successive queries with the same parameters (and hence cache key, such as the default search) continuously made faster than the cache expiration time (the more likely due to multiple users in production), then the cache never expires?

Reply:
... or I suppose at least one client must encounter AffectingTimeSpan being invalid, and that clears the cache, though.

Reply:
It's not the cache itself that dictates whether the results are invalid. There is a `valid_unit` parameter attached to each cached result, which is checked on each request to validate a specific value. Then even if one of those values is invalid, the whole cache is recalculated.

Reply:
There is but the `valid_until` value is set again, to the same value for the previous page items as the items from the current page which are newly added to the cache (if I'm reading it correctly)?
Even if so, it may not matter in practical terms.

Reply:
Nothing else pops out. Nice tests.